enhancements/installer/byo-custom-dns.md

```diff
@@ -0,0 +1,573 @@
+- Start with AWS and lay the groundwork to add this capability for other cloud
+platforms like Azure and GCP.
+- Continue using the default cloud LB service for API, API-Int and Ingress.
+
+### Non-Goals
+
+We will not use the [Kubernetes External DNS provider](https://github.com/openshift/external-dns)
+and/or the [OpenShift External DNS Operator](https://github.com/openshift/external-dns-operator)
+to program dns entries across varied DNS providers. The External DNS Operator
+relies on being able to program the customer's DNS solution. The customer DNS
+solutions in consideration here do not allow OpenShift to configure them.
+
+## Proposal
+
+For the OpenShift installation to make use of a custom DNS solution that is not
+configurable by OpenShift, the customer is responsible for configuring the DNS
+entries for the API, API-Int and Ingress services.
+
+Currently, during cluster installation, OpenShift is repsonsible for adding
+entries for API, API-Int and *.apps into the cloud DNS. These URLs are
+configured with the LB DNS Names of the LBs associated with each of these
+services. Hence the LBs have to be created before they can used to configure
+the DNS solution. Also, important to note is that these LB DNS Names cannot
+be generated before they are created.
+
+When it is indicated to the OpenShift installer via an input parameter that the
+custom DNS solution is in use, OpenShift does not configure the cloud DNS
+solution [Route53 for AWS](https://aws.amazon.com/route53/) for any of its
+services. In the case of API and API-Int services, the customer is responsible
+for creating the external and internal LBs [Elastic Load Balancer](https://aws.amazon.com/elasticloadbalancing/)
+prior to cluster installation. Their custom DNS solution also needs to be
+configured with entries for API and API-Int before cluster installation. OpenShift
+will continue to configure and manage these cloud LBs created by the user.
+
+DNS configuration for Ingress services will be handled differently. The Ingress
+Operator utilizes the upstream k8s LoadBalancer service to create and configure
+the LBs for the *.apps service. In this case, the customer pre-creating the LB
+would mean the LB service would have to be modified to consume a pre-created LB.
+Accomplishing this would require a change to the upstream LB service potentially
+adding delays towards the completion of this feature. Since the LB DNS Name for
+the *.apps would not be available before cluster installation, the DNS solution
+cannot be pre-configured before cluster installation either.
+
+To work around this limitation, the Ingress Operator would continue using the
+LoadBalancer service to create and configure the LBs for the *.apps service.
+OpenShift will start an in-cluster DNS pod to provide DNS resolution for *.apps
+during cluster installation. After the cluster is installed, the user will have
+to update their custom DNS solution with an entry for the *.apps service.
```

Comment:
I didn't know that assisted-installer already did this. We would have to do the same with non-assisted installs too when `userDNS` is `Enabled` on supported platforms.

Reply:
I believe Ingress Operator also has a canary check.

Reply:
Sandhya tried this and the [results](https://prow.ci.openshift.org/view/gs/origin-ci-test/pr-logs/pull/openshift_installer/7403/pull-ci-openshift-installer-master-e2e-aws-ovn/1688663966525427712) were consistent with that - ingress, auth, and console operators failed.

It's tempting to think about how we could change that, but if we're having to add DNS entries for api and api-int anyway, then there's no harm in adding *.apps as well.
In theory it might be possible to handle api and api-int (but not the wildcard) with /etc/hosts instead of running a full DNS; that would be the only reason to think about this any further.